os/kernel/pthread/pthread_keydestroy.c

```diff
@@ -0,0 +1,105 @@
+/****************************************************************************
+ *
+ * Copyright 2018 Samsung Electronics All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND,
+ * either express or implied. See the License for the specific
+ * language governing permissions and limitations under the License.
+ *
+ ****************************************************************************/
+
+/************************************************************************
+ * Included Files
+ ************************************************************************/
+
+#include <stdlib.h>
+#include <sys/types.h>
+#include <queue.h>
+
+#include <tinyara/sched.h>
+#include <tinyara/kmalloc.h>
+
+/************************************************************************
+ * Definitions
+ ************************************************************************/
+
+/************************************************************************
+ * Private Type Declarations
+ ************************************************************************/
+
+/************************************************************************
+ * Global Variables
+ ************************************************************************/
+
+/************************************************************************
+ * Private Variables
+ ************************************************************************/
+
+/************************************************************************
+ * Private Functions
+ ************************************************************************/
+
+/************************************************************************
+ * Public Functions
+ ************************************************************************/
+
+/************************************************************************
+ * Name: pthread_key_destroy
+ *
+ * Description:
+ *   When pthread is finished by exit or cancel, this function executes
+ *   key destructor and releases key resources.
+ *
+ * Parameters:
+ *   None
+ *
+ * Returned Value:
+ *   None
+ *
+ * Assumptions:
+ *
+ ************************************************************************/
+
+void pthread_key_destroy(struct pthread_tcb_s *tcb)
+{
+	struct task_group_s *group = tcb->cmn.group;
+	struct pthread_key_s *cur_key;
+	struct pthread_key_s *next_key;
+	int destr_count = 0;
+
+	for (cur_key = (struct pthread_key_s *)sq_peek(&tcb->key_list); cur_key; cur_key = next_key) {
+
+		/* Execute destructor until exceeding limitation */
+
+		if (destr_count < PTHREAD_DESTRUCTOR_ITERATIONS) {
```

Comment:
That is written in POSIX.
```
If, after all the destructors have been called for all non-NULL values with associated destructors,
there are still some non-NULL values with associated destructors, then the process is repeated.
If, after at least {PTHREAD_DESTRUCTOR_ITERATIONS} iterations of destructor calls for outstanding non-NULL values,
there are still some non-NULL values with associated destructors, implementations may stop calling destructors,
or they may continue calling destructors until no non-NULL values with associated destructors exist,
even though this might result in an infinite loop.
```